File: copymove/src/main/scala/farjs/copymove/CopyMovePlugin.scala
package farjs.copymove

import farjs.copymove.CopyMoveUiAction._
import farjs.filelist._
import farjs.filelist.api.{FileListCapability, FileListItem}
import farjs.filelist.stack.{PanelStack, WithStacksProps}
import scommons.react.ReactClass
import scommons.react.blessed.BlessedElement

import scala.scalajs.js

object CopyMovePlugin extends FileListPlugin(js.Array("f5", "f6", "S-f5", "S-f6")) {

  override def onKeyTrigger(key: String,
                            stacks: WithStacksProps,
                            data: js.UndefOr[js.Dynamic] = js.undefined): js.Promise[js.UndefOr[ReactClass]] = {
    
    val (maybeFrom, maybeTo, toInput) =
      if (stacks.left.stack.isActive) {
        (getData(stacks.left.stack), getData(stacks.right.stack), stacks.right.input)
      }
      else (getData(stacks.right.stack), getData(stacks.left.stack), stacks.left.input)

    val res = key match {
      case "f5" | "f6" =>
        maybeFrom.zip(maybeTo).flatMap { case (from, to) =>
          onCopyMove(key == "f6", from, to, toInput).map { action =>
            new CopyMoveUi(action, from, Some(to)).apply()
          }
        }
      case "S-f5" | "S-f6" =>
        maybeFrom.flatMap { from =>
          onCopyMoveInplace(key == "S-f6", from).map { action =>
            new CopyMoveUi(action, from, None).apply()
          }
        }
      case _ => None
    }

    js.Promise.resolve[js.UndefOr[ReactClass]](res match {
      case Some(r) => r
      case None => js.undefined
    })
  }

  private def getData(stack: PanelStack): Option[FileListData] = {
    val item = stack.peek[js.Any]()
    item.getData().toOption
  }

  private[copymove] final def onCopyMoveInplace(move: Boolean, from: FileListData): Option[CopyMoveUiAction] = {
    FileListState.currentItem(from.state).filter(_ != FileListItem.up).toOption.flatMap { _ =>
      if (move && from.actions.api.capabilities.contains(FileListCapability.moveInplace)) {
        Some(ShowMoveInplace)
      }
      else if (!move && from.actions.api.capabilities.contains(FileListCapability.copyInplace)) {
        Some(ShowCopyInplace)
      }
      else None
    }
  }

  private[copymove] final def onCopyMove(move: Boolean,
                                         from: FileListData,
                                         to: FileListData,
                                         toInput: BlessedElement): Option[CopyMoveUiAction] = {

    val currItem = FileListState.currentItem(from.state).filter(_ != FileListItem.up)

    if ((from.state.selectedNames.nonEmpty || currItem.nonEmpty) &&
      from.actions.api.capabilities.contains(FileListCapability.read) &&
      (!move || from.actions.api.capabilities.contains(FileListCapability.delete))) {

      if (to.actions.api.capabilities.contains(FileListCapability.write)) {
        if (move) Some(ShowMoveToTarget)
        else Some(ShowCopyToTarget)
      }
      else {
        toInput.emit("keypress", js.undefined, js.Dynamic.literal(
          name = "",
          full =
            if (move) FileListEvent.onFileListMove
            else FileListEvent.onFileListCopy
        ))
        None
      }
    }
    else None
  }
}

File: copymove/CopyMovePlugin.mjs
/**
 * @typedef {import("@farjs/blessed").Widgets.BlessedElement} BlessedElement
 * @import { FileListData } from "@farjs/filelist/FileListData.mjs"
 * @import { CopyMoveUiAction } from "./CopyMoveUi.mjs"
 */
import FileListCapability from "@farjs/filelist/api/FileListCapability.mjs";
import FileListItem from "@farjs/filelist/api/FileListItem.mjs";
import PanelStack from "@farjs/filelist/stack/PanelStack.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";
import FileListPlugin from "@farjs/filelist/FileListPlugin.mjs";
import FileListEvent from "@farjs/filelist/FileListEvent.mjs";
import CopyMoveUi from "./CopyMoveUi.mjs";

class CopyMovePlugin extends FileListPlugin {
  constructor() {
    super(["f5", "f6", "S-f5", "S-f6"]);
  }

  /** @type {FileListPlugin['onKeyTrigger']} */
  async onKeyTrigger(key, stacks) {
    /** @type {(stack: PanelStack) => FileListData | undefined } */
    const getData = (stack) => stack.peek().getData();

    const [from, maybeTo, toInput] = stacks.left.stack.isActive
      ? [
          getData(stacks.left.stack),
          getData(stacks.right.stack),
          stacks.right.input,
        ]
      : [
          getData(stacks.right.stack),
          getData(stacks.left.stack),
          stacks.left.input,
        ];

    switch (key) {
      case "f5":
      case "f6":
        if (from && maybeTo) {
          const action = this._onCopyMove(key === "f6", from, maybeTo, toInput);
          return action
            ? CopyMoveUi({ show: action, from, maybeTo })
            : undefined;
        }
        return undefined;
      case "S-f5":
      case "S-f6":
        if (from) {
          const action = this._onCopyMoveInplace(key === "S-f6", from);
          return action ? CopyMoveUi({ show: action, from }) : undefined;
        }
        return undefined;
    }

    return undefined;
  }

  /**
   * @param {boolean} move
   * @param {FileListData} from
   * @returns {CopyMoveUiAction | undefined}
   */
  _onCopyMoveInplace(move, from) {
    if (FileListState.currentItem(from.state, (i) => i !== FileListItem.up)) {
      if (
        move &&
        from.actions.api.capabilities.has(FileListCapability.moveInplace)
      ) {
        return "ShowMoveInplace";
      }
      if (
        !move &&
        from.actions.api.capabilities.has(FileListCapability.copyInplace)
      ) {
        return "ShowCopyInplace";
      }
    }

    return undefined;
  }

  /**
   * @param {boolean} move
   * @param {FileListData} from
   * @param {FileListData} to
   * @param {BlessedElement} toInput
   * @returns {CopyMoveUiAction | undefined}
   */
  _onCopyMove(move, from, to, toInput) {
    const currItem = FileListState.currentItem(
      from.state,
      (i) => i !== FileListItem.up,
    );

    if (
      (from.state.selectedNames.size > 0 || currItem) &&
      from.actions.api.capabilities.has(FileListCapability.read) &&
      (!move || from.actions.api.capabilities.has(FileListCapability.delete))
    ) {
      if (to.actions.api.capabilities.has(FileListCapability.write)) {
        return move ? "ShowMoveToTarget" : "ShowCopyToTarget";
      }

      toInput.emit("keypress", undefined, {
        name: "",
        full: move
          ? FileListEvent.onFileListMove
          : FileListEvent.onFileListCopy,
      });
    }

    return undefined;
  }
}

CopyMovePlugin.instance = new CopyMovePlugin();

export default CopyMovePlugin;
